src/main/java/backend/RepoIO.java

```diff
@@ -51,14 +52,32 @@ public RepoIO(boolean isTestMode, boolean enableTestJSON) {
 
     public CompletableFuture<Model> openRepository(String repoId) {
         if (repoStore.isRepoStored(repoId)) {
-            return repoStore.loadRepository(repoId)
-                .thenCompose(this::updateModel)
-                .exceptionally(withResult(new Model(repoId)));
+            return loadRepositoryFromStore(repoId);
         } else {
-            return repoSource.downloadRepository(repoId)
+            return downloadRepositoryFromSource(repoId);
+        }
+    }
+
+    private CompletableFuture<Model> loadRepositoryFromStore(String repoId) {
+        return repoStore.loadRepository(repoId)
+                .thenCompose(this::updateModel)
+                .exceptionally(ex -> {
+                    try {
```

Comment:
I actually struggle with quite a number of dilemmas for this fix since every solution seems to be either ugly in their own way or add more complications than I feel they deserve it. It's not really a problem with the chaining thing, but it's the lack of expressiveness in the current java's functional interface.

I wanted to do the handling in `openRepository` at first since that makes the most naturally sense. However, we have to implement part of the *Futures* chain in `loadRepo...` and the other part in `openRepository` since the chain returns immediately and may crash at arbitrary time, so a surrounding `try/catch` is technically useless. And it seems unsymmetrical for me b/w the `downloadRepo...` and `loadRepo...` (aka there is a mismatch in level of abstraction). When a developer uses these 2 functions, `downloadRepo...` handles everything nicely in a package while `loadRepo...` requires its user to handle exceptions separately. What makes it worse is that I cannot make `loadRepo...` *throws exceptions* loudly because the exceptions are produced within the monad chain's mechanism. So I justify that `loadRepo...` is handling something weird (that `RepoStore.isRepoStored` returns `true` but the repo is technically not stored) and should, as part of its behavior, handle this case behide the scene, and to the viewpoint of a user of this method, it's behaving as expected (well, except for being slower). This behavior could be made explicit by a docs for this method. We can also make things symmetrical by taking `exceptionally` out of both `downloadRepo...` and `loadRepo...` but then there is, uuh, DRY and everything :P. Anw, actually I don't have a very strong opinion on this one since all of the solutions that I think could work are ugly in some ways, so a vote or any althernative suggestion?

Now, the `try/catch` looks weird in the flow of the chain and we could try to generalize this like `withResult`. But then it seems that in this case, the exception is modifying the control flow in a quite dangerous way with the chain from `downloadRepo...` having deep side-effects (`updateModel`) and blocking. A generic exception handler would not be aware of the uncertain nature of its possibly multiple argument CompletableFutures and could be more restricting than making things more flexible (I guess that's why java doesn't allow thenm in `exceptionally` in the first place). So I prefer not to handle these *exceptions* in a generic way but feel more comfortable having all the *try/catch/throw/exception* smacked onto my face loudly where they belong :P. Having such a method would be a good structural convenience though, if we have more snippets that handle exceptions by having 1 wrapped value, 1 unwrapped value and a default error message with trivial execution footprints then it would be a good time to refactor.

Then a nicely packaged way is to extend java's functional CompletableFutures to be more expressive, but then I feel it does not yet deserve the additional complications given our purpose.